chainer/optimizers/adam.py

```diff
@@ -240,12 +240,18 @@ def update_core_gpu(self, param):
                     'T eps, T eta, T weight_decay_rate, raw T dummy',
                     'P param, P m, P v, P vhat',
                     '''T grad_ = static_cast<T>(grad);
-                       m += one_minus_beta1 * (grad_ - m);
-                       v += one_minus_beta2 * (grad_ * grad - v);
-                       vhat = max(vhat, v);
+                       T m_ = static_cast<T>(m);
+                       T v_ = static_cast<T>(v);
+                       T vhat_ = static_cast<T>(vhat);
+                       m_ += one_minus_beta1 * (grad_ - m_);
+                       v_ += one_minus_beta2 * (grad_ * grad_ - v_);
+                       vhat_ = max(vhat_, v_);
+                       vhat = static_cast<T>(vhat_);
+                       m = static_cast<T>(m_);
+                       v = static_cast<T>(v_);
```

Comment:
`vhat`, `m` and `v` have the type `P`. What's the point of those casts?

Reply:
Sorry, typo. Should be casted back to P.
Im retrying the checks to see if it fails or not.
If everything goes well I will push the update later.